src/cargo/ops/cargo_fetch.rs

```diff
@@ -1,12 +1,102 @@
-use core::{PackageSet, Resolve, Workspace};
+use core::compiler::{BuildConfig, Kind, TargetInfo};
+use core::{Package, PackageIdSpec, PackageSet, Resolve, Workspace};
 use ops;
+use std::collections::HashSet;
 use util::CargoResult;
+use util::Config;
+
+pub struct FetchOptions<'a> {
+    pub config: &'a Config,
+    /// The target arch triple to fetch dependencies for
+    pub target: Option<String>,
+}
 
 /// Executes `cargo fetch`.
-pub fn fetch<'a>(ws: &Workspace<'a>) -> CargoResult<(Resolve, PackageSet<'a>)> {
+pub fn fetch<'a>(
+    ws: &Workspace<'a>,
+    options: &FetchOptions<'a>,
+) -> CargoResult<(Resolve, PackageSet<'a>)> {
     let (packages, resolve) = ops::resolve_ws(ws)?;
-    for id in resolve.iter() {
-        packages.get(id)?;
+
+    if let Some(_) = options.target {
+        fetch_for_target(ws, options.config, &options.target, &resolve, &packages)?;
+    } else {
+        for id in resolve.iter() {
+            packages.get(id)?;
+        }
     }
+
     Ok((resolve, packages))
 }
+
+fn fetch_for_target<'a, 'cfg: 'a>(
+    ws: &'a Workspace<'cfg>,
+    config: &'cfg Config,
+    target: &Option<String>,
+    resolve: &Resolve,
+    packages: &'a PackageSet<'cfg>,
+) -> CargoResult<HashSet<&'a Package>> {
+    let mut set = HashSet::new();
+    let jobs = Some(1);
+    let build_config = BuildConfig::new(config, jobs, target, None)?;
+    let target_info = TargetInfo::new(config, &build_config, Kind::Target)?;
+    let root_pkgs = ws.default_members()
+        .map(Package::package_id)
+        .map(PackageIdSpec::from_package_id)
```

Comment:
I think you can skip the `PackageIdSpec` step here and go straight to `packages.get` after the initial `package_id` above, right?